agent/config/file_watcher.go

```diff
@@ -0,0 +1,221 @@
+
+	w := &FileWatcher{watcher: ws,
+		configFiles:      cfgFiles,
+		handleFunc:       handleFunc,
+		reconcileTimeout: timeoutDuration,
+		done:             make(chan interface{}),
+	}
+	w.logger = logger.Named("file-watcher")
+	for _, f := range configFiles {
+		err = w.add(f)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return w, nil
+}
+
+func (w *FileWatcher) Start(ctx context.Context) {
+	cancelCtx, cancel := context.WithCancel(ctx)
+	w.cancel = cancel
+	go w.watch(cancelCtx)
+}
+
+func (w *FileWatcher) add(filename string) error {
+	if isSymLink(filename) {
+		return fmt.Errorf("symbolic link are not supported %s", filename)
+	}
+	filename = filepath.Clean(filename)
+	w.logger.Trace("adding file", "file", filename)
+	if err := w.watcher.Add(filename); err != nil {
+		return err
+	}
+	iNode, err := w.getFileId(filename)
+	if err != nil {
+		return err
+	}
+	w.configFiles[filename] = &watchedFile{id: iNode}
+	return nil
+}
+
+func isSymLink(filename string) bool {
+	fi, err := os.Lstat(filename)
+	if err != nil {
+		return false
+	}
+	if fi.Mode()&os.ModeSymlink != 0 {
+		return true
+	}
+	return false
+}
+
+func (w *FileWatcher) Stop() error {
+	w.cancel()
+	<-w.done
+	return w.watcher.Close()
+}
+
+func (w *FileWatcher) watch(ctx context.Context) {
+	ticker := time.NewTicker(w.reconcileTimeout)
+	defer ticker.Stop()
+	for {
+		select {
+		case event, ok := <-w.watcher.Events:
+			if !ok {
+				w.logger.Error("watcher event channel is closed")
+				return
+			}
+			w.logger.Debug("received watcher event", "event", event)
+			if err := w.handleEvent(event); err != nil {
+				w.logger.Error("error handling watcher event", "error", err, "event", event)
+			}
+		case _, ok := <-w.watcher.Errors:
+			if !ok {
+				w.logger.Error("watcher error channel is closed")
+				return
+			}
+		case <-ticker.C:
+			w.reconcile()
+		case <-ctx.Done():
+			close(w.done)
+			return
+		}
+	}
+
+}
+
+func (w *FileWatcher) handleEvent(event fsnotify.Event) error {
+	w.logger.Debug("event received ", "filename", event.Name, "OP", event.Op)
+	// we only want Create and Remove events to avoid triggering a relaod on file modification
+	if !isCreate(event) && !isRemove(event) && !isWrite(event) && !isRename(event) {
+		return nil
+	}
+	filename := filepath.Clean(event.Name)
+	configFile, basename, ok := w.isWatched(filename)
+	if !ok {
+		return fmt.Errorf("file %s is not watched", event.Name)
+	}
+
+	// we only want to update inode and re-add if the event is on the watched file itself
+	if filename == basename {
+		if isRemove(event) {
+			// If the file was removed, try to reconcile and see if anything changed.
+			w.logger.Debug("attempt a reconcile ", "filename", event.Name, "OP", event.Op)
+			configFile.id = 0
+			w.reconcile()
+		}
+	}
+	if isCreate(event) || isWrite(event) || isRename(event) {
+		w.logger.Debug("call the handler", "filename", event.Name, "OP", event.Op)
+		go w.handleFunc(&WatcherEvent{Filename: filename})
```

Comment:
Why spawn a new goroutine for every one of these events that trigger a reload? You may have multiple racing reloads in play if you do that.

It might be better to have a buffered handoff channel that you write to here, and a single goroutine that serially reloads parts of the config as they change.

Reply:
yes, that's a good point. I think I will change the API to have `NewFileWatcher` return a channel and the user can read the events from that channel.